src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Editor.Xaml;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Xaml.Diagnostics.Analyzers;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Roslyn.Utilities;
+using Shell = Microsoft.VisualStudio.Shell;
+
+namespace Microsoft.VisualStudio.LanguageServices.Xaml
+{
+    [Export]
+    internal sealed partial class XamlProjectService
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly Workspace _workspace;
+        private readonly VisualStudioProjectFactory _visualStudioProjectFactory;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactory;
+        private readonly IThreadingContext _threadingContext;
+        private readonly Shell.RunningDocumentTable _rdt;
+        private readonly IVsSolution _vsSolution;
+        private uint? _rdtEventsCookie;
+        private readonly Dictionary<IVsHierarchy, VisualStudioProject> _xamlProjects = new();
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public XamlProjectService(
+            [Import(typeof(Shell.SVsServiceProvider))] IServiceProvider serviceProvider,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            VisualStudioProjectFactory visualStudioProjectFactory,
+            VisualStudioWorkspace workspace,
+            IXamlDocumentAnalyzerService analyzerService,
+            IThreadingContext threadingContext)
+        {
+            _serviceProvider = serviceProvider;
+            _editorAdaptersFactory = editorAdaptersFactoryService;
+            _visualStudioProjectFactory = visualStudioProjectFactory;
+            _workspace = workspace;
+            _threadingContext = threadingContext;
+            _rdt = new Shell.RunningDocumentTable(_serviceProvider);
+            _vsSolution = (IVsSolution)_serviceProvider.GetService(typeof(SVsSolution));
+            _vsSolution.AdviseSolutionEvents(this, out _);
```

Comment:
Do any of these operations require the UI thread?  If so it shouldn't be done in the MEF constructor

Reply:
I simply moved it. This has been in the MEF constructor for XamlTextViewCreationListener before for years without issues so it should be fine in its new location as well.

Reply:
I don't know all the quirks of vs mef initialization, but the manner of initialization here is changing slightly.  It can now be triggered by the ILanguageClient initialization instead of text view initialization which could affect correctness (perhaps text view initialization was always run on the UI thread).  tagging @sharwell for thoughts here

Reply:
I just checked in the debugger. The XamlProjectService constructor does get called on the UI thread in the Nexus client scenario.